### PR TITLE
Forward auto-grading settings to DL assignments

### PR DESCRIPTION
**Pull request by marcospri**
For:

- https://github.com/hypothesis/lms/issues/6647

### Deep linking summary:

- The frontend calls the deep_linking_fields API with the assignment's config.
- The BE returns a JWT that must be POSTed to the LMS client side.
- That JWT contains an encoded representation of the assigned configuration. Some of those elements are encoded in and URL other are port of the launch's custom parameters.
- The client POSTs the config the LMS, the assignment is created 
- Next launches of the assignments forward to us the configuration.

This PR handles encoding the auto grading configuration as part of the deep linked parameters.

Next steps will need to read it on launches and persist it to the DB.

### Testing

- Requires: https://github.com/hypothesis/lms/pull/6632
- Configure a LTI1.3 assignment with autograding, any parameters.
- Save the config on the canvas side
- Apply a diff like:

```diff
diff --git a/lms/views/lti/basic_launch.py b/lms/views/lti/basic_launch.py
index b2404a85b..c40859098 100644
--- a/lms/views/lti/basic_launch.py
+++ b/lms/views/lti/basic_launch.py
@@ -81,6 +81,11 @@ class BasicLaunchViews:
                     request=self.request, type_=LTIEvent.Type.CONFIGURED_LAUNCH
                 )
             )
+            print(
+                self.request.lti_params.v13.get(
+                    "https://purl.imsglobal.org/spec/lti/claim/custom"
+                )
+            )
             return {}
 
         # Show the file-picker for the user to choose a document.
```

- Launch the assignment (example done in https://hypothesis.instructure.com/courses/319/assignments/7296)
- You'll see the auto grading settings as part of the launch parameters.

### CODE REVIEW

File: lms/views/lti/deep_linking.py [hunk 1]
```diff
@@ -85,12 +86,32 @@ def deep_linking_launch(context, request):
     return {}
 
 
+class _AutoGradingConfig(Schema):
+    grading_type = fields.Str(
+        required=True, validate=validate.OneOf(["all_or_nothing", "scaled"])
+    )
+    activity_calculation = fields.Str(
+        required=True, validate=validate.OneOf(["cumulative", "separately"])
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:94]: "combined"?

File: lms/views/lti/deep_linking.py [hunk 2]
```diff
@@ -85,12 +86,32 @@ def deep_linking_launch(context, request):
...
+    )
+
+    required_annotations = fields.Int(
+        required=False, allow_none=True, validate=validate.Range(min=0)
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:98]: min=1?

File: lms/views/lti/deep_linking.py [hunk 3]
```diff
@@ -85,12 +86,30 @@ def deep_linking_launch(context, request):
     return {}
 
 
+class _AutoGradingConfig(Schema):
+    grading_type = fields.Str(
+        required=True, validate=validate.OneOf(["all_or_nothing", "scaled"])
+    )
+    activity_calculation = fields.Str(
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:93]: Changed this to  "separate"

File: lms/views/lti/deep_linking.py [hunk 4]
```diff
@@ -85,12 +86,30 @@ def deep_linking_launch(context, request):
...
+        required=True, validate=validate.OneOf(["cumulative", "separate"])
+    )
+
+    required_annotations = fields.Int(required=True, validate=validate.Range(min=0))
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:97]: required_annotations always required

File: tests/unit/lms/views/lti/deep_linking_test.py [hunk 5]
```diff
@@ -211,6 +211,10 @@ def test_it_for_v11(
                 {"group_set": "1", "title": "title"},
                 {"group_set": "1", "title": "title"},
             ),
+            (
+                {"auto_grading_config": {"key": "value"}},
+                {"auto_grading_config": '{"key": "value"}'},
```
Comment (acelaya) [inline tests/unit/lms/views/lti/deep_linking_test.py:216]: Are we not validating that this fulfills `_AutoGradingConfig` definition? 🤔 

I would expect these values to be invalid.

**Review by acelaya — COMMENTED** [review]
I noticed the tests are only covering the usage of invalid values, but in that case I would expect the test to fail.

File: tests/unit/lms/views/lti/deep_linking_test.py (hunk 5 above)
Comment (marcospri) [reply tests/unit/lms/views/lti/deep_linking_test.py:275]: Yes, good point but I'm afraid I don't a satisfactory answer.

The method here is unaware of the types, it just passes along what it receives, and in the case of these settings it serlizes it as json. That's what the test does. I included the private name of the method in the test name (one extra `_`)

The marshmallow schema is validated before we get into the the code of individual views and it gets serialized into `request.parsed_params`, if that not valid the view is never called.

That being said I added a very broad valid/invalid tests for the schema itself :+1:

File: tests/unit/lms/views/lti/deep_linking_test.py [hunk 6]
```diff
@@ -5,13 +5,72 @@
...
+from lms.validation import ValidationError
+from lms.views.lti.deep_linking import (
+    DeepLinkingFieldsRequestSchema,
+    DeepLinkingFieldsViews,
+    deep_linking_launch,
+)
 from tests import factories
 
 
+class TestDeepLinkingFieldsRequestSchema:
```
Comment (marcospri) [inline tests/unit/lms/views/lti/deep_linking_test.py:21]: Testing just the schema here.

File: tests/unit/lms/views/lti/deep_linking_test.py [hunk 7]
```diff
@@ -211,9 +270,13 @@ def test_it_for_v11(
...
                 {"group_set": "1", "title": "title"},
             ),
+            (
+                {"auto_grading_config": {"key": "value"}},
+                {"auto_grading_config": '{"key": "value"}'},
+            ),
         ],
     )
-    def test_get_assignment_configuration(
+    def test__get_assignment_configuration(
```
Comment (marcospri) [inline tests/unit/lms/views/lti/deep_linking_test.py:279]: Fixed this test's name, it testing `_get_assignment_configuration` not `get_assignment_configuration``

File: tests/unit/lms/views/lti/deep_linking_test.py (hunk 5 above)
Comment (acelaya) [reply tests/unit/lms/views/lti/deep_linking_test.py:275]: > The marshmallow schema is validated before we get into the the code of individual views and it gets serialized into `request.parsed_params`, if that not valid the view is never called.

Oh, I see 👍🏼

**Review by acelaya — APPROVED** [review]

**Pull request merged**